pytensor/tensor/rewriting/subtensor.py

```diff
@@ -381,6 +381,51 @@ def local_useless_slice(fgraph, node):
             return [node.inputs[0]]
 
 
+@register_infer_shape
+@register_canonicalize
+@register_specialize
+@node_rewriter([Subtensor])
+def local_replace_slice(fgraph, node):
+    """
+    Remove Subtensor of the form:
+        1. X[0:-1:1] -> X[:-1]
+        2. X[0:-1] -> X[:-1]
+        3. X[:-1] -> X[:-1]
+
+    """
+    idxs = get_idx_list(node.inputs, node.op.idx_list)
+
+    if not idxs:
+        return [node.inputs[0]]
+
+    # last_slice = len(idxs)
+
+    for s in idxs[::-1]:
+        if (
+            isinstance(s, slice)
+            and (
+                s.start is None
+                or extract_constant(s.start, only_process_constants=True) == 0
+            )
+            and extract_constant(s.stop, only_process_constants=True) == -1
+            and (
+                s.step is None
+                or extract_constant(s.step, only_process_constants=True) == 1
+            )
+        ):
+            # This does not work.
+            # I get the error that
+            # ```
+            # s.start = None
+            # AttributeError: readonly attribute
+            # ```
+            s.start = None
+            s.stop = -1
+            s.step = None
```

Comment:
This does not work.
I get the error that
 ```
s.start = None
 AttributeError: readonly attribute
```
I think that making a new slice object with start, stop and step being None, -1, None might be the solution. But I am not sure about how to do that so need help regarding this.

Also all I want to do is:
```
DeepCopyOp [id H] ''   5
 |Subtensor{int64:int64:int64} [id I] ''   4
   |x [id C]
   |ScalarConstant{0} [id G]
   |ScalarConstant{-1} [id D]
   |ScalarConstant{1} [id J]
   ```
   to something like:
   ```
   DeepCopyOp [id A] ''   1
 |Subtensor{:int64:} [id B] ''   0
   |x [id C]
   |ScalarConstant{-1} [id D]
   ```
   so if you think of a better way to achieve this. please suggest :)

Reply:
Yes you can't overwrite the slice, you need to define a new one with `slice(new_start, new_stop, new_step)` which may or not all be new.

Also the stop does not matter, that was just the example. The useless stop would be `x[:, len(stop)]` which we can get rid of if x.type.shape is constant across that dimension and is equal to the constant stop.

Your logic is a bit too mixed up now. I suggest structuring the function to go through the indexes and:
0. Convert the index tuple to a list so you can overwrite it's entries
0. Set a boolean flag indicating if any slice should be updated. Start with `False`
1. Then iterate over the indexes
2. Check if it's a slice. If not continue
3. Check if start is a constant zero, in which case it should be replaced by None, otherwise left as is
4. Check if stop is a constant with the same value as x.type.shape along that dimension, in which case it should be replaced by None
5. Check that step is a constant 1, in which case it should be replaced by None
6. If anything should be changed, create a new slice with with the new (or potential same old) start, step, stop. Overwrite the list of indexes with that new slice and set the boolean flag to true

After finishing the iteration over all the indexes: check if the boolean flag is true. In that case you should return [x[tuple(idxs)]], otherwise you should return None.

This rewrite can potentially be combined with `local_useless_slice` which removes trailing `None` slices